README.md
# zlog
[![Gem Version](https://badge.fury.io/rb/zlog.png)](http://badge.fury.io/rb/zlog)
[![Build Status](https://travis-ci.org/arlimus/zlog.png)](https://travis-ci.org/arlimus/zlog)

Logging configuration on top of ruby's `logging` gem (see [github](https://github.com/TwP/logging)).

* colorful default logging to stdout
* supports continuous logging (eg progress indicators which you don't want to clutter your commandline)
* added section and ok log types

## requirements

* gems:
  * logging

## installation

From rubygems:

    gem install zlog

From source:

    gem build *.gemspec && gem install *gem

## example

Code:

    require "zlog"
    Zlog.init_stdout loglevel: :debug

    l = Logging.logger["main"]
    l.section "log demonstration"
    l.ok "ok me"
    l.debug "debug me"
    l.info "info me"
    l.warn "warn me"
    l.error "error me"
    l.fatal "fatal me"


See the `example` folder for more.

## utility

You can use `zlog` to read and format json-based log files:

    cat file.log | zlog

It will print out a formatted representation of the log.

![Example image in example/example.output.png](https://raw.github.com/arlimus/zlog/master/example/example.output.png)


CHANGELOG.md
## 0.9.2

* bugfix: corrected handling of jsony-messages in cli tool

## 0.9.1

* bugfix: handle multiple log messages in json

## 0.9.0

* bugfix: allow continuous readinf from stdin, without waiting for eof

## 0.8

* feature: support reading log files for zlog formatter
* feature: added convertor from json (log file format) to logevent in zlog
* feature: get a writable logfile from a list of candidates or generate a random one

## 0.7

* improvement: correctly ordered log levels (now: ... warn ok section error ...)
* bugfix: don't set root log level when initializing stdout logger; make sure logging to other appenders is not affected here
* bugfix: prevent duplicate entries of stdout appender + prevent removal of other appenders when initializing zlog stdout

## 0.6

* improvement: fixed versions of gem file dependencies
* bugfix: deprecated ruby 1.9.2

## 0.5

* improvement: added test suite with rake, minitest, spec, turn and travis

## 0.4

* feature: added abort( message, exit_code ) to call a fatal log message and exit
* improvement: added changelog
* bugfix: removed highline as dependency, it is not necessary anymore

## 0.3

* feature: switched to logging framework for ruby; this is now a formatter/styler on top of it
* bugfix: removed prefix zeros from 256-colors ansi calls to blue and green

## 0.2

* improvement: made debug logging slightly lighter to be more readable
* bugfix: removed color-mode detection, using 256-colors by default now

## 0.1

* feature: initial stable release


test/minitest_helper.rb
require 'minitest/autorun'
require 'zlog'

begin
  require 'turn'
  Turn.config do |c|
   # use one of output formats:
   # :outline  - turn's original case/test outline mode [default]
   # :progress - indicates progress with progress bar
   # :dotted   - test/unit's traditional dot-progress mode
   # :pretty   - new pretty reporter
   # :marshal  - dump output as YAML (normal run mode only)
   # :cue      - interactive testing
   c.format  = :pretty
   # turn on invoke/execute tracing, enable full backtrace
   # c.trace   = true
   # use humanized test names (works only with :outline format)
   c.natural = true
  end
rescue LoadError
end

test/classes/zlog_spec.rb
require 'minitest_helper'

describe Zlog do
  it "can initialize itself to STDOUT" do
    Zlog.must_respond_to :init_stdout
  end

  describe 'get_writable_logfile' do

    before :each do
      @random_writable = File::join( Dir::tmpdir, Dir::Tmpname.make_tmpname( "", nil ) )
      @random_non_writable = File::join( '/dev', Dir::Tmpname.make_tmpname("",nil) )
    end

    it "must give a working writable log file" do
      Zlog.get_writable_logfile( @random_writable ).
        must_equal @random_writable
    end

    it "must give return the first actually writable log file" do
      Zlog.get_writable_logfile( @random_non_writable, @random_writable ).
        must_equal @random_writable
    end

    it "must provide a random file if none of the provided candidates are writable" do
      f = Zlog.get_writable_logfile( @random_non_writable, @random_writable )
      f.must_be_instance_of String
      f.wont_be_empty
    end

  end

  describe 'json_2_event' do
    before :each do
      template = '{"timestamp":"2013-10-10T18:37:38.513438+02:00","level":"%level","logger":"Mocker%level","message":"mock message %level"}'
      names = ["DEBUG","INFO","WARN","OK","SECTION","ERROR","FATAL"]
      @examples = (0..(names.length-1)).
        map do |level|
          {
            raw: template.gsub( '%level', names[level] ),
            level: level,
            name: names[level]
          }
        end
    end

    it "must give nil on incorrect json" do
      Zlog.json_2_event( "" ).must_be_nil
    end

    it "must convert a logging json to a logevent" do
      @examples.each do |e|
        r = Zlog.json_2_event e[:raw]
        r.must_be_instance_of Logging::LogEvent
        r.level.must_equal e[:level]
        r.logger.must_equal "Mocker#{e[:name]}"
        r.data.must_equal "mock message #{e[:name]}"
      end
    end

  end
end


test/classes/logger_spec.rb
require 'minitest_helper'

describe Logging::Logger do
  before :each do
    @log = Logging.logger["main"]
  end

  it "provides a method for logging 'ok' messages" do
    @log.must_respond_to :ok
  end

  it "provides a method for logging 'section's" do
    @log.must_respond_to :section
  end

  it "provides a method for logging 'abort's" do
    @log.must_respond_to :abort
  end

  it "provides a method for logging 'cont'inuous messages" do
    @log.must_respond_to :cont
  end
end


example/example.rb
#!/bin/ruby
require_relative '../lib/zlog'

Zlog.init_stdout loglevel: :debug

logger = Logging.logger["zlog"]
logger.section "zlog"
logger.ok "ok me"
logger.debug "debug me"
logger.info "info me"
logger.warn "warn me"
logger.error "error me"
logger.fatal "fatal me"

logger.section "continuous logging"
(1..50).each do |len|
  logger.debug( logger.cont "o"*len+"k" )
  sleep 0.01
end
logger.info "back to normal"

logger.section "logging with hierarchies"
Zlog.init_stdout named: true, loglevel: :debug
foo = Logging.logger['Foo']
bar = Logging.logger['Foo::Bar']
baz = Logging.logger['Foo::Baz']
foo.debug 'foo will send a warning soon'
foo.warn 'this is a warning, not a ticket'
bar.info 'this message will not be logged'
baz.info 'nor will this message'
bar.ok 'bar ís ok'
bar.error 'but this error message will be logged'

logger.section "show logger configuration"
Logging.show_configuration

lib/zlog/simple_layouts.rb

module Zlog::Layouts::SimpleCore
  def format_simple event, print_name = false, pallette = STDOUT_PATTERN_256COLORS
    # Logging::LogEvent logger="zlog", level=2, data="warn me", time=2013-09-29 23:36:12 +0200, file="", line="", method=""
    level = LOGLEVEL2NAME[event.level]
    pattern = pallette[level]
    # get the core message
    mkmsg = lambda{ |data|
        obj = format_obj(data)
        if print_name
          (pallette[:name] % event.logger) + ( pattern % obj ) 
        else 
          pattern % obj 
        end
      }
    # handle continuous logging lines
    if (event.data.start_with? "\r")
      # this is a continuous message
      # there are 2 cases which need to be covered:
      # A: the last line was a regular log line,
      #    the cursor caret is at position 0
      # B: the last line was a continuous log line,
      #    the cursor caret is at the end of the last message
      # For case B we want to go back to the beginning of the line (via "\r")
      # then insert the message and the amount of white space needed to
      # overwrite any visible characters from the last message
      msg = mkmsg.(event.data[1..-1])
      # determine the length of the last message
      len = (@has_last_line_newline == false) ? @last.length : 0
      # calculate the amount of white spaces we need to overwrite
      # remnants of the last log message
      rem_len = len - msg.length
      rem_len = 0 if rem_len < 0
      # create the resulting string
      ret = "\r" + msg + (' '*rem_len)
      # make sure to update state parameters
      @last = msg
      @has_last_line_newline = false
      # return the result
      ret
    else
      # format the object
      msg = mkmsg.(event.data)
      # check if we need to add a newline at the start of the line
      # this only happens when the last message was a continuous message
      # and thus didn't set a \n at the end
      sl =
        if (@has_last_line_newline == false)
          @has_last_line_newline = true
          "\n"
        else "" end
      # return the resulting pattern
      sl + msg + "\n"
    end
  end

  # standard order: ["DEBUG", "INFO", "WARN", "ERROR", "FATAL"]

  LOGLEVEL2NAME = [:debug, :info, :warning, :ok, :section, :error, :fatal]

  STDOUT_PATTERN_NOCOLORS = {
    :debug    => ".. %s",
    :info     => "-- %s",
    :warning  => "ww %s",
    :ok       => "++ %s",
    :section  => "\n== %s",
    :error    => "ee %s",
    :fatal    => "ff %s",
    :name     => "%s: "
  }

  STDOUT_PATTERN_8COLORS = {
    :debug   => "\033[37m.. %s\033[0m",
    :info    => "\033[0m-- %s\033[0m",
    :warning => "\033[1;33mww %s\033[0m",
    :ok      => "\033[1;32m++ %s\033[0m",
    :section => "\n\033[1;34m== %s\033[0m",
    :error   => "\033[1;31mee %s\033[0m",
    :fatal   => "\033[1;41mff\033[0m\033[1;31 %s\033[0m",
    :name    => "\033[37m%s: "
  }

  STDOUT_PATTERN_256COLORS = {
    :debug   => "\033[38;5;246m.. %s\033[0m",
    :info    => "\033[38;5;255m-- %s\033[0m",
    :warning => "\033[38;5;226mww %s\033[0m",
    :ok      => "\033[38;5;46m++ %s\033[0m",
    :section => "\n\033[38;5;33m== %s\033[0m",
    :error   => "\033[38;5;196mee %s\033[0m",
    :fatal   => "\033[48;5;196mff\033[0m\033[38;5;196m %s\033[0m",
    :name    => "\033[38;5;246m%s: "
  }

  STDOUT_SYMBOLIC_PATTERN_256COLORS = {
    :debug   => "\033[38;5;246m. %s\033[0m",
    :info    => "\033[38;5;255m━ %s\033[0m",
    :warning => "\033[38;5;226m! %s\033[0m",
    :ok      => "\033[38;5;46m✔ %s\033[0m",
    :section => "\n\033[38;5;33m= %s\033[0m",
    :error   => "\033[38;5;196m✘ %s\033[0m",
    :fatal   => "\033[48;5;196m✘\033[0m\033[38;5;196m %s\033[0m",
    :name    => "\033[38;5;246m%s: "
  }
end

module Zlog::Layouts
  # Accessor / Factory for Simple layout
  def self.simple( *args )
    ::Zlog::Layouts::Simple.new(*args)
  end

  def self.named( *args )
    ::Zlog::Layouts::SimpleNamed.new(*args)
  end

  def self.symbolic( *args )
    ::Zlog::Layouts::SimpleSymbolic.new(*args)
  end

  # Simple layout for easy readability
  class Simple < ::Logging::Layout
    include Zlog::Layouts::SimpleCore
    # format log events
    def format( event )
      format_simple event, false
    end
  end

  # Simple layout for easy readability
  class SimpleNamed < ::Logging::Layout
    include Zlog::Layouts::SimpleCore
    # format log events
    def format( event )
      format_simple event, true
    end
  end

  # Simple layout for easy readability
  class SimpleSymbolic < ::Logging::Layout
    include Zlog::Layouts::SimpleCore
    # format log events
    def format( event )
      format_simple event, false, Zlog::Layouts::SimpleCore::STDOUT_SYMBOLIC_PATTERN_256COLORS
    end
  end
end


lib/zlog/zlog.rb
require 'json'
require 'tmpdir'

module Zlog
  VERSION = "0.9.2"
  extend self

  def init_stdout opts = {layout: :simple, loglevel: nil}
    # determine the layout from options
    layout = case opts[:layout]
      when :named
        Zlog::Layouts.named
      else
        Zlog::Layouts.simple
      end

    # configure the default stdout appender
    Logging.appenders.stdout(
      level: opts[:loglevel],
      layout: layout
    )
    # find all non-stdout appenders
    # this prevents duplicates of stdout appender
    as = Logging.logger.root.appenders.find_all{|i| i.name != "stdout" }
    # add back all appenders + our own stdout appender
    Logging.logger.root.appenders = as + ["stdout"]
  end

  def get_writable_logfile *candidates
    ret = candidates.compact.find do |f|
      # testing if something is writable, options:
      # #1: discarged, returns false if the file doesn't exist;
      #     it would require to check the folder too... :(
      # File::writable? f
      #2 trial+error
      begin
        File.open(f,"a").close
        true
      rescue
        false
      end
    end
    # in case none of the provided logfiles is writable, create one
    if ret.nil?
      # try to get a prefix as the filename from the first provided log file
      prefix = File::basename( candidates[0].to_s )
      # determine a random filename
      File::join( Dir::tmpdir, Dir::Tmpname.make_tmpname( prefix, nil ) )
    else
      # return the found good filename
      ret
    end
  end

  def json_2_event str
    begin
      j = JSON::load str
      # LogEvent = Struct.new( :logger, :level, :data, :time, :file, :line, :method )
      # from: https://github.com/TwP/logging/blob/master/lib/logging/log_event.rb
      # json example:
      # {"timestamp":"2013-10-10T18:37:38.513438+02:00","level":"DEBUG",
      #  "logger":"Log1-debug","message":"debugging info, this won't show up on console"}
      Logging::LogEvent.new( j['logger'], Logging.level_num(j['level']), j['message'].to_s, j['timestamp'] )
    rescue
      nil
    end
  end
end
